sycl/source/detail/error_handling/enqueue_kernel.cpp

```diff
@@ -48,62 +71,64 @@ bool oclHandleInvalidWorkGroupSize(const device_impl &DeviceImpl,
-  } else {
-    // OpenCL 2.x:
-    // PI_INVALID_WORK_GROUP_SIZE if local_work_size is specified and the
-    // total number of work-items in the work-group computed as
-    // local_work_size[0] * ... * local_work_size[work_dim – 1] is greater
-    // than the value specified by PI_KERNEL_GROUP_INFO_WORK_GROUP_SIZE in
-    // table 5.21.
-    size_t KernelWGSize = 0;
-    Plugin.call<PiApiKind::piKernelGetGroupInfo>(
-        Kernel, Device, PI_KERNEL_GROUP_INFO_WORK_GROUP_SIZE, sizeof(size_t),
-        &KernelWGSize, nullptr);
-    const size_t TotalNumberOfWIs =
-        NDRDesc.LocalSize[0] * NDRDesc.LocalSize[1] * NDRDesc.LocalSize[2];
-    if (TotalNumberOfWIs > KernelWGSize)
-      throw sycl::nd_range_error(
-          "Total number of work-items in a work-group cannot exceed "
-          "info::kernel_work_group::work_group_size which is equal to " +
-              std::to_string(KernelWGSize) + " for this kernel",
-          PI_INVALID_WORK_GROUP_SIZE);
+  if (Platform.get_backend() == cl::sycl::backend::opencl) {
+    if (Ver[0] == '1') {
+      // OpenCL 1.x:
+      // PI_INVALID_WORK_GROUP_SIZE if local_work_size is specified and the
+      // total number of work-items in the work-group computed as
+      // local_work_size[0] * ... * local_work_size[work_dim – 1] is greater
+      // than the value specified by PI_DEVICE_MAX_WORK_GROUP_SIZE in
+      // table 4.3
+      size_t MaxWGSize = 0;
+      Plugin.call<PiApiKind::piDeviceGetInfo>(
+          Device, PI_DEVICE_INFO_MAX_WORK_GROUP_SIZE, sizeof(size_t),
+          &MaxWGSize, nullptr);
+      const size_t TotalNumberOfWIs =
+          NDRDesc.LocalSize[0] * NDRDesc.LocalSize[1] * NDRDesc.LocalSize[2];
+      if (TotalNumberOfWIs > MaxWGSize)
+        throw sycl::nd_range_error(
+            "Total number of work-items in a work-group cannot exceed "
+            "info::device::max_work_group_size which is equal to " +
+                std::to_string(MaxWGSize),
+            PI_INVALID_WORK_GROUP_SIZE);
+    } else {
+      // RELEVENT // OpenCL 2.x:
+      // PI_INVALID_WORK_GROUP_SIZE if local_work_size is specified and the
+      // total number of work-items in the work-group computed as
+      // local_work_size[0] * ... * local_work_size[work_dim – 1] is greater
+      // than the value specified by PI_KERNEL_GROUP_INFO_WORK_GROUP_SIZE in
+      // table 5.21.
+      size_t KernelWGSize = 0;
+      Plugin.call<PiApiKind::piKernelGetGroupInfo>(
+          Kernel, Device, PI_KERNEL_GROUP_INFO_WORK_GROUP_SIZE, sizeof(size_t),
+          &KernelWGSize, nullptr);
+      const size_t TotalNumberOfWIs =
+          NDRDesc.LocalSize[0] * NDRDesc.LocalSize[1] * NDRDesc.LocalSize[2];
+      if (TotalNumberOfWIs > KernelWGSize)
+        throw sycl::nd_range_error(
+            "Total number of work-items in a work-group cannot exceed "
+            "info::kernel_work_group::work_group_size which is equal to " +
+                std::to_string(KernelWGSize) + " for this kernel",
```

Comment:
"for this kernel" doesn't help if I have more than one. I suggest adding a kernel name to the message.

Reply:
Note, however, that:

* Kernels do not always have a name that is meaningful to the user since the `KernelName` template parameter is optional as of SYCL 2020 provisional.

* The exception is thrown synchronously when submitting a kernel, so the problematic kernel should be clear if the application uses try/catch around the code that submits the kernel.

Adding the name of the kernel to the exception string does seem reasonable in the case where the user specified a name.  If we decide to do that, we should do it uniformly for all the exceptions diagnosed in this function, not just for this one exception.